backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java

```diff
@@ -642,25 +645,30 @@ private SyncToParentStatus deriveSyncToParentStatus(
 
       String remoteTrackingBranchName = branch.getRemoteTrackingBranch().map(rtb -> rtb.getName()).getOrNull();
 
-      Function<IGitCoreCommitHash, Seq<String>> getRelevantContainingBranches = commitHash -> deriveBranchesContainingGivenCommitInReflog()
-          .getOrElse(commitHash, List.empty())
-          .filter(candidateBranchName -> !candidateBranchName.equals(branch.getName())
-              && !candidateBranchName.equals(remoteTrackingBranchName)
-              // We demand that the candidate branch is already managed.
-              && managedBranchNames.contains(candidateBranchName));
-
-      IGitCoreCommit upstreamIndicator = gitCoreRepository.findFirstSatisfyingAncestor(branch.derivePointedCommit(),
-          commitHash -> getRelevantContainingBranches.apply(commitHash).nonEmpty()).getOrNull();
+      var commitAndContainingBranches = gitCoreRepository
+          .ancestorsOf(branch.derivePointedCommit())
+          .map(commit -> {
+            var containingManagedBranches = deriveBranchesContainingGivenCommitInReflog()
+                .getOrElse(commit.getHash(), List.empty())
+                .filter(candidateBranchName -> !candidateBranchName.equals(branch.getName())
+                    && !candidateBranchName.equals(remoteTrackingBranchName)
+            // We demand that the candidate branch is already managed.
```

Comment:
can we have it indented as the lines below and above?

Reply:
This will be reworked soon on PR for discover anyway